### PR TITLE
@types/hapi - v17

### PR DESCRIPTION
**Pull request by rafaelsouzaf**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/hapijs/hapi/blob/master/API.md
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by gitawego** [comment]
congrats!!!

**Comment by rafaelsouzaf** [comment]
Thanks @gitawego ! Unfortunatelly Travis is not  _hapi_ LOL.

File: types/hapi/definitions/request/request-events.d.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    /**
+     * Access: read only and the public podium interface.
+     * The request.events supports the following events:
+     * * 'peek' - emitted for each chunk of payload data read from the client connection. The event method signature is function(chunk, encoding).
+     * * 'finish' - emitted when the request payload finished reading. The event method signature is function ().
+     * * 'disconnect' - emitted when a request errors or aborts unexpectedly.
+     * [See docs](https://github.com/hapijs/hapi/blob/master/API.md#-requestevents)
+     */
+    on(criteria: RequestEventType, listener: Function): void;
```
Comment (SimonSchick) [inline types/hapi/definitions/request/request-events.d.ts]: You can use a mapped type to declare all `on/once` methods on this interface.

File: types/hapi/definitions/server/server-method.d.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * The method function with a signature async function(...args, [flags]) where:
+ * * ...args - the method function arguments (can be any number of arguments or none).
+ * * flags - when caching is enabled, an object used to set optional method result flags:
+ * * * ttl - 0 if result is valid but cannot be cached. Defaults to cache policy.
+ * For reference [See docs](https://github.com/hapijs/hapi/blob/master/API.md#-servermethodname-method-options)
+ */
+export interface ServerMethod {
+//    (args: any[], flags: {ttl?: number}): any;
+    (...args: any[]): any; // TODO it needs review. Look the doc: the first param is a ...args and the second is "flags". Typescript supports ...args only in the last param
```
Comment (SimonSchick) [inline types/hapi/definitions/server/server-method.d.ts]: This should probably be a type, not an interface?

File: types/hapi/definitions/server/server-route.d.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * route custom rules object. The object is passed to each rules processor registered with server.rules(). Cannot be used if route.options.rules is defined.
+     */
+    rules?: object;
+
+    /**
+     * TODO: Missing documentation. Exist only in examples and test files.
+     */
+    config?: any;
```
Comment (SimonSchick) [inline types/hapi/definitions/server/server-route.d.ts]: Probably want to make this a consumer extensible interface.

File: types/hapi/definitions/server/server-state-options.d.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+     *  * password - password used for HMAC key generation (must be at least 32 characters long).
+     */
+    sign?: {
+        integrity?: object;  // TODO make iron definitions and getting typing from iron. Needs review!
+        password: string;
+    };
+    /** password used for 'iron' encoding (must be at least 32 characters long). */
+    password?: string;
+    /** options for 'iron' encoding. Defaults to require('iron').defaults. */
+    iron?: object;  // TODO make iron definitions and getting typing from iron. Needs review!
```
Comment (SimonSchick) [inline types/hapi/definitions/server/server-state-options.d.ts]:
```
declare module 'iron' {
    export interface ISomething {
        saltBits: number;
        algorithm: string;
        iterations: number;
        minPasswordlength: number;
    }

    export interface ISealOptions {
        encryption: ISomething;
        integrity: ISomething;

        ttl: number;
        timestampSkewSec: number;
        localtimeOffsetMsec: number;
    }

    export const defaults: ISealOptions;

    export const algorithms: {
        'aes-128-ctr': {
            keyBits: number;
            ivBits: number;
        };
        'aes-256-cbc': {
            keyBits: number;
            ivBits: number;
        };
        'sha256': {
            keyBits: number;
        };
    };

    export const macFormatVersion: string;
    export const macPrefix: string;

    export interface IGenerateKeyOptions extends Pick<ISomething, 'algorithm' | 'iterations' | 'minPasswordlength'> {
        saltBits?: number;
        salt?: string;
        iv?: string;
    }

    export interface IKey {
        key: Buffer;
        salt: string;
        iv: string;
    }

    export function generateKey (password: string, options: IGenerateKeyOptions): Promise<IKey>;
    export function encrypt (password: string, options: IGenerateKeyOptions, data: string): Promise<{ data: Buffer, key: IKey }>;
    export function decrypt (password: string, options: IGenerateKeyOptions, data: string): Promise<Buffer>;

    export interface IHMacResult {
        digest: string;
        salt: string;
    }

    export function hmacWithPassword(password: string, options: IGenerateKeyOptions, data: string): Promise<IHMacResult>;

    export function seal (obj: object, password: string, options: ISealOptions): Promise<string>;
    export function unseal (data: string, password: string, options: ISealOptions): Promise<object>;
}
```

no warranties for correctness.

File: types/hapi/test/server/server-cache-provision.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    port: 8000,
+});
+server.initialize();
+server.cache.provision({engine: require('catbox-memory'), name: 'countries' });
+
+const cache:catbox.Policy = server.cache({segment: 'countries', cache: 'countries', expiresIn: 60 * 60 * 1000 });
+cache.set('norway', 'oslo', 10 * 1000, null);
+const value = cache.get('norway', null);
+
+// TODO Not working. It needs review.
```
Comment (SimonSchick) [inline types/hapi/test/server/server-cache-provision.ts]: ?

File: types/hapi/test/server/server-inject.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    handler: (request: Request, h: ResponseToolkit) => {
+        return 'Success!';
+    }
+};
+
+server.route(serverRoute);
+server.start();
+
+const res = server.inject('/');
+console.log(res.result);                // 'Success!' // TODO It's not working
```
Comment (SimonSchick) [inline types/hapi/test/server/server-inject.ts]: ??

File: types/hapi/definitions/util/lifecycle.d.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    export type ReturnValue = ReturnValueTypes | (Promise<ReturnValueTypes>);
+    export type ReturnValueTypes =
+        (null | undefined | string | number | boolean) |
+        (Buffer) |
+        (Error | Boom.BoomError) |
+        (stream.Stream) |
+        (Object | Object[]) |
+        // a toolkit signal:
+        // a toolkit method response:
+        any; // TODO need review
```
Comment (SimonSchick) [inline types/hapi/definitions/util/lifecycle.d.ts]: Probably remove `any` in favour of stricter types?

**Review by SimonSchick — CHANGES_REQUESTED** [review]

File: types/hapi/definitions/util/lifecycle.d.ts (hunk 7 above)
Comment (rafaelsouzaf) [reply types/hapi/definitions/util/lifecycle.d.ts]: Commit 244e1ee. Replaced "any" with other types after reviewing the documents again.

File: types/hapi/test/server/server-cache-provision.ts (hunk 5 above)
Comment (rafaelsouzaf) [reply types/hapi/test/server/server-cache-provision.ts]: Sorry. It was an old comment. @BorntraegerMarc told me to take it out but I forgot. Commit 3fe936d

File: types/hapi/test/server/server-inject.ts (hunk 6 above)
Comment (rafaelsouzaf) [reply types/hapi/test/server/server-inject.ts]: The same here. I forgot to remove this comment. Commit 3fe936d

File: types/hapi/definitions/server/server-state-options.d.ts (hunk 4 above)
Comment (rafaelsouzaf) [reply types/hapi/definitions/server/server-state-options.d.ts]: Nice! Thanks @SimonSchick. Is it your code or did you get it from another place?

If it's new, and looks good comparing to [Iron js](https://github.com/hueniverse/iron/blob/master/lib/index.js) code, is not better to make a new PR creating the ```/types/iron``` ? Or should I put this code inside our hapi definitions?

File: types/hapi/definitions/server/server-route.d.ts (hunk 3 above)
Comment (rafaelsouzaf) [reply types/hapi/definitions/server/server-route.d.ts]: @SimonSchick Yes, I understand. The problem is that I don't have any docs or a good example to make a interface of this config object. For that I preferred to use an "any" for now. But I am open for new ideas of course :-)

File: types/hapi/definitions/server/server-method.d.ts (hunk 2 above)
Comment (rafaelsouzaf) [reply types/hapi/definitions/server/server-method.d.ts]: Commit b923c71. Thanks!

File: types/hapi/definitions/request/request-events.d.ts (hunk 1 above)
Comment (rafaelsouzaf) [reply types/hapi/definitions/request/request-events.d.ts]: @SimonSchick can you give me more info please? I read about Mapped Type ([link-1](https://www.typescriptlang.org/docs/handbook/advanced-types.html), [link-2](https://blog.mariusschulz.com/2017/01/20/typescript-2-1-mapped-types)) but I can't see how to use in this case. We are overriding methods (on/once) from Podium interface but with some differents parameters. How can I use Mapped Type in this case please? Thanks for the feedback!

File: types/hapi/definitions/server/server-state-options.d.ts (hunk 4 above)
Comment (SimonSchick) [reply types/hapi/definitions/server/server-state-options.d.ts]: I just figured I would post it here, I didn't have time to PR these typings yet, feel free to PR them yourself.

Comment (SimonSchick) [reply types/hapi/definitions/server/server-state-options.d.ts]: And yes, I wrote those typings.

File: types/hapi/definitions/request/request-events.d.ts (hunk 1 above)
Comment (SimonSchick) [reply types/hapi/definitions/request/request-events.d.ts]: https://github.com/SimonSchick/EliteDangerousUtils/blob/master/src/EDLog/EDLog.ts#L307

File: types/hapi/definitions/server/server-route.d.ts (hunk 3 above)
Comment (SimonSchick) [reply types/hapi/definitions/server/server-route.d.ts]: I'd just make an empty interface and export it.

Comment (rafaelsouzaf) [reply types/hapi/definitions/server/server-route.d.ts]: Commit 065b070

File: types/hapi/definitions/server/server-state-options.d.ts (hunk 4 above)
Comment (rafaelsouzaf) [reply types/hapi/definitions/server/server-state-options.d.ts]: Commit 62ae76a. Thanks @SimonSchick

**Comment by SimonSchick** [comment]
@aozgaa this PR was not ready for merge...

**Comment by rafaelsouzaf** [comment]
Tomorrow I will start working again with this PR 👍 (or made a new one with updates).

**Comment by SimonSchick** [comment]
I just fail to understand why this was merged without any approvals...

@andy-ms @aozgaa

**Comment by rafaelsouzaf** [comment]
@SimonSchick Please I need to push some commits. Should I create another PR (from another local branch)?

**Comment by SimonSchick** [comment]
@rafaelsouzaf The PR was reverted, you should be able to push the branch again and keep commiting, or open a new PR.

**Comment by rafaelsouzaf** [comment]
@SimonSchick now I have a big mess here man :-/
The new commits are not visible here more. Probably because the status is "merged". And I have a huge conflict. I will try to fix it today and create a new PR.

**Comment by andy-ms** [comment]
@rafaelsouzaf Could you just check out your branch as it was, copy the directory, then checkout master and move it back? I don't think we need to preserve all the separate commits from this branch.

**Comment by rafaelsouzaf** [comment]
@andy-ms No problem. I made a new PR [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/22462). Thanks!

@SimonSchick Well. I will keep working on this new PR. But in this moment Travis is complaining because the iron library is set with TS 2.6, and hapi v17 with 2.4:
`Error: hapi depends on iron but has a lower required TypeScript version.`

Can I downgrade the iron to 2.4? Thanks!

**Comment by SimonSchick** [comment]
Hapi 17 was released post TS2.6, it makes sense to me to force them all to ts 2.6, but I guess downgrading is fine too.

**Comment by rafaelsouzaf** [comment]
@SimonSchick I agree and I tried before. But if I use TS 2.6 in Hapi17, I will have similar error with all this types because they are using 2.4 or 2.2:

```
    /types/joi
    /types/boom
    /types/shot
    /types/mimos
    /types/podium
    /types/catbox
    /types/mime-db
```

The last one "mime-db" is not a "hapi project". I am not sure about the consequences in changing all this types.

Maybe I can submit a PR to downgrade the iron version now and go ahead with the PR.

**Pull request merged**